New Luis Recognizer

emadelwany opened the pull request:
DO NOT MERGE => Need to publish the Microsoft.Cognitive.LUIS v3.0.0 package to nuget first

cleemullins commented on Microsoft.Bot.Builder.LUIS.Tests/LuisRecognizerTests.cs:
```diff
@@ -0,0 +1,170 @@
+        private static IRecognizer GetLuisRecognizer(bool verbose = false, ILuisOptions luisOptions = null)
+        {
+            var luisRecognizerOptions = new LuisRecognizerOptions {Verbose = verbose};
+            var luisModel = new LuisModel("6209a76f-e836-413b-ba92-a5772d1b2087", "f2eef6a1cab345b9b4b53743357e869f", new Uri("https://westus.api.cognitive.microsoft.com/luis/v2.0/apps/"), LuisApiVersion.V2);
```
This... isn't allowed. No checking keys into github. 

You'll now have to depriate / rotate the password. Security teams continually sweep for this stuff (even in GitHub history and branches) and will send you a nasty mail.

cleemullins replied:
Why is the URI tied to westus? Wouldn't the DNS name that's one of the Anycast / GeoDNS names be better?

emadelwany replied:
Moving these values to environment variables (similar to how other recognizers are being tested). 
LUIS doesn't have a global DNS (something to do with how keys are provisioned per region), specific region must be specified.

emadelwany commented:
This implementation is a port of the NodeJS implementation:
https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botbuilder-ai/src/luisRecognizer.ts

cleemullins commented on Microsoft.Bot.Builder.LUIS.Tests/LuisRecognizerTests.cs:
```diff
@@ -0,0 +1,170 @@
+using Microsoft.Cognitive.LUIS;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Bot.Builder.LUIS.Tests
```
I'm not clear how these tests could be run on the Build Machine. 

Please look at how we're doing this in the [other tests](https://github.com/Microsoft/botbuilder-dotnet/blob/master/tests/Microsoft.Bot.Builder.Ai.Tests/QnAMakerTests.cs#L15). And the solution [we're using](https://github.com/Microsoft/botbuilder-dotnet/blob/d28d218f78a735f6b43399844b04b3bc654fda23/tests/Microsoft.Bot.Builder.Tests/TestUtilities.cs#L40).

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+
+        private async Task<RecognizerResult> Recognize(LuisRequest request, CancellationToken ct, bool verbose)
+        {
+            var luisResult = await _luisService.QueryAsync(request, ct);
```
.ConfigurareAwait(false);

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+            return recognizerResult;
+        }
+
+        private static JObject GetIntents(LuisResult luisResult)
```
Consider simplifying this code, so it's not doing so many things on a single line.

emadelwany replied:
I think it's easy to follow as-is, and VS now supports breakpoints inside line so debugging it is easy too. I find this functional style easier to follow them imperative code and looping when there isn't really business logic but just transformations of data.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+
+        private static JObject GetIntents(LuisResult luisResult)
+        {
+            return luisResult.Intents != null ?
```
Consider Simplifying.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                entitiesAndMetadata["$instance"] = new JObject();
+            }
+            var compositeEntityTypes = new HashSet<string>();
```
Why a HashSet? List<string> seems like it would do the trick, and be a little more clear. 

(Semantics are the same, but anytime I see HashSet, I wonder "Why did they pick this?". In this case, it doesn't seem specific to the problem domain, so just List<T> would be more clear...)

emadelwany replied:
HashSet to make look ups fast (O(1)), some LUIS models have hundreds of entities.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+            return Regex.Replace(entity.Type, "\\.", "_");
+        }
+
+        private static IList<EntityRecommendation> PopulateCompositeEntity(CompositeEntity compositeEntity, IList<EntityRecommendation> entities, JObject entitiesAndMetadata, bool verbose)
```
Consider breaking up into sub-mehods.

emadelwany replied:
I made the method shorter, I think it would make it harder to follow if I break it any further.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizerMiddleware.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public LuisRecognizerMiddleware(ILuisModel luisModel, ILuisRecognizerOptions luisRecognizerOptions = null, ILuisOptions luisOptions = null)
+        {
+            _luisRecognizer = new LuisRecognizer(luisModel, luisRecognizerOptions, luisOptions);
```
Null-check the luisMidel and throw if it's null.

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizerMiddleware.cs:
```diff
@@ -0,0 +1,37 @@
+            BotAssert.ContextNotNull(context);
+
+            var utterance = context.Request.AsMessageActivity()?.Text;
+            var result = await _luisRecognizer.Recognize(utterance, CancellationToken.None).ConfigureAwait(false);
```
For all non-messages (ConversationJoin, etc) this will result in the utterance being null. 

That, in turn, will result int he luisRecognizer.Recognizer being called with a null utterance. That's probably not needed.

Consider scoping this while thing to just run if the Request.Type == ActivityTypes.Message

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/ILuisRecognizerOptions.cs:
```diff
@@ -0,0 +1,16 @@
+        /// <summary>
+        /// If set to true, metadata is added to the recognizer's results
+        /// </summary>
+        bool Verbose { get; }
```
What about Spell Checking and the other features that folks have asked for? 

Wouldn't this come from the LUIS NuGet package?

emadelwany replied:
These are the "Recognizer Options, what you're talking about (LUIS Options) are exposed in ILuisOptions

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisModel.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    public class LuisModel : ILuisModel
+    {
+        public LuisModel(string modelId, string subscriptionKey, Uri uriBase, LuisApiVersion apiVersion)
```
Null checking + throw on all of these paremeters.

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisModel.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public LuisRequest ModifyRequest(LuisRequest request)
+        {
+            return request;
```
I think I'm missing something obvious. What's the point of this method?

emadelwany replied:
I'm working around the annoying way the bot builder v3 Luis Client is implemented, didn't want invest in changing the v3 implementation so much since it's been very stable.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisModel.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public double Threshold => 0.0d;
+
+        public LuisRequest ModifyRequest(LuisRequest request)
```
null-check + throw

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+        private readonly ILuisOptions _luisOptions;
+        private readonly ILuisRecognizerOptions _luisRecognizerOptions;
+
+        public LuisRecognizer(ILuisModel luisModel, ILuisRecognizerOptions luisRecognizerOptions = null, ILuisOptions options = null)
```
null-check + throw on luisModel

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+        public LuisRecognizer(ILuisModel luisModel, ILuisRecognizerOptions luisRecognizerOptions = null, ILuisOptions options = null)
+        {
+            _luisService = new LuisService(luisModel);
+            _luisOptions = options;
```
If not options is passed in (null), consider creating a "Default" option set and setting that here. 

That'll get rid of all the null-checks + special cases in your code later on. It also gives a clear place to inspect / change the default options.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+        {
+            _luisService = new LuisService(luisModel);
+            _luisOptions = options;
+            _luisRecognizerOptions = luisRecognizerOptions;
```
Same comment around create a default set of options if none is passed in. Then remove all the null-checks + special cases in the code below.

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+        }
+
+        /// <inheritdoc />
+        public Task<RecognizerResult> Recognize(string utterance, CancellationToken ct)
```
null-check + throw on the utterance. 

if (string.IsNullOrEmpty(utterance)) throw new ARgumentNullExceptions(nameof(utterance));

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+        {
+            var luisRequest = new LuisRequest(utterance);
+            _luisOptions?.Apply(luisRequest);
+            return Recognize(luisRequest, ct, _luisRecognizerOptions?.Verbose ?? true);
```
Special cases like this will go away if you use the default option approach. That'll also get rid of all the null-coalescing operators which make code hard to write, less clear, and more bug prone.

emadelwany replied:
Done

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+                for (var i = 0; i < entities.Count; i++)
+                {
+                    var entity = entities[i];
+                    if (!coveredSet.Contains(i) &&
```
If statements like this are too complex. Please simplify.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -0,0 +1,217 @@
+            }
+
+            // filter entities that were covered by this composite entity
+            for (var i = 0; i < entities.Count; i++)
```
This seems weird. Is this code right? Why wouldn't this just be a standard for-each?

cleemullins replied:
As it sits, myself (and everyone looking at it) wonders, "Is it entities.Count, or entities.Count- 1?"

emadelwany replied:
Will refactor this

cleemullins commented:
@gabog, take a look and see if this is heading down the right path for your needs.

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs:
```diff
@@ -22,24 +23,30 @@ public class LuisRecognizer : IRecognizer
 
+        private readonly ILuisOptions _defaultLuisOptions = new LuisRequest();
+        private readonly ILuisRecognizerOptions _defaultRecognizerOptions = new LuisRecognizerOptions {Verbose = true};
+
```
Why have both a _luisOptions and a _defaultLuisOptions? 

Just initialize:
```
_luisOptions = new LuisRequest();
_luisRecognizerOptions = new LuisRecognizerOptions {Verbose = true};
_luisOptions = new LuisRequest();
_luisRecognizerOptions = new LuisRecognizerOptions {Verbose = true};
```
... and call it done.

emadelwany replied:
I already refactored this but didn't push, did something different but similar

cleemullins commented on libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizerMiddleware.cs:
```diff
@@ -20,17 +22,22 @@ public class LuisRecognizerMiddleware : IReceiveActivity
-            context.Set(LuisRecognizerResultKey, result);
+            if (context.Request.Type == ActivityTypes.Message)
+            {
+                var utterance = context.Request.AsMessageActivity()?.Text;
```
You don't need the "?" any more.

emadelwany replied:
Will you promise? :-)

cleemullins commented on tests/Microsoft.Bot.Builder.LUIS.Tests/LuisRecognizerTests.cs:
```diff
@@ -17,17 +19,29 @@ namespace Microsoft.Bot.Builder.LUIS.Tests
         [TestMethod]
         public async Task SingleIntent_SimplyEntity()
         {
+            if (_luisAppId == null || _subscriptionKey == null || _luisUriBase == null)
```
This if statment shows up in several places. Consider making a method.

emadelwany replied:
Done

emadelwany commented:
@gabog I think this PR addresses all your feedback:
1. You can now toggle "verbose" LUIS option and get all intents (not just top)
2. You can pass all other options to LUIS (spell check, logging etc)
3. You can access values of entities in a consistent way.

I recommend that you look at the unit tests for a high level overview of the new schema and how you access intents and entities.

emadelwany commented:
@cleemullins the LUIS Client nuget will take a few days to actually publish (due to policies and processes out of our control), so I added the code (as-is copy) to this repo for now, and I'll remove it and replace usages with references to the nuget once it's published, if that seems ok to you I'd like to merge this for now.

cleemullins commented:
I'm good with that to get Gabo (and others) unblocked.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________
From: Emad Elwany <notifications@github.com>
Sent: Wednesday, March 7, 2018 10:48 PM
To: Microsoft/botbuilder-dotnet
Cc: Chris Mullins; Mention
Subject: Re: [Microsoft/botbuilder-dotnet] New Luis Recognizer (#228)


@cleemullins<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcleemullins&data=04%7C01%7Ccmullins%40microsoft.com%7C711bf805752c45606efd08d584c0a1de%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636560885271235068%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-2&sdata=QKm3rPLSJsEOAdGTdw0t0l6P3OTs6nt6AXrwd5fskrU%3D&reserved=0> the LUIS Client nuget will take a few days to actually publish (due to policies and processes out of our control), so I added the code (as-is copy) to this repo for now, and I'll remove it and replace usages with references to the nuget once it's published, if that seems ok to you I'd like to merge this for now.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fbotbuilder-dotnet%2Fpull%2F228%23issuecomment-371396404&data=04%7C01%7Ccmullins%40microsoft.com%7C711bf805752c45606efd08d584c0a1de%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636560885271235068%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-2&sdata=7sTN%2BbM7l4juizi2oZYfNfsMw4%2B%2FJIN0kuucKORcNZ8%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABHICahWVfFCAAJCWV7A8XsrCtK3AV9gks5tcNRJgaJpZM4SY_NP&data=04%7C01%7Ccmullins%40microsoft.com%7C711bf805752c45606efd08d584c0a1de%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636560885271235068%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-2&sdata=eH22Wtzn3NKTikH7fQrw2GXuphlHGENBPLJlBLcxTCE%3D&reserved=0>.

cleemullins commented:
This has been merged. Let's get it fully working! :) 

@gabog: Progress!

bebenekmarcin commented:
@emadelwany when you are planning to publish the Microsoft.Cognitive.LUIS v3.0.0 package to nuget?
